src/render_draw_bonus.c

```diff
@@ -0,0 +1,90 @@
+/* ************************************************************************** */
+/*                                                                            */
+/*                                                        :::      ::::::::   */
+/*   render_draw_bonus.c                                :+:      :+:    :+:   */
+/*                                                    +:+ +:+         +:+     */
+/*   By: sqiu <sqiu@student.42vienna.com>           +#+  +:+       +#+        */
+/*                                                +#+#+#+#+#+   +#+           */
+/*   Created: 2023/11/20 15:52:55 by gwolf             #+#    #+#             */
+/*   Updated: 2024/01/14 19:29:05 by sqiu             ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "render_bonus.h"
+
+static uint32_t	ft_fast_alpha_blend(uint32_t bg_color, t_menu menu)
+{
+	uint32_t	bg_rb;
+	uint32_t	bg_g;
+	uint32_t	blend_rb;
+	uint32_t	blend_g;
+
+	bg_rb = (bg_color & 0xFF00FF) * menu.inv_alpha;
+	bg_g = (bg_color & 0x00FF00) * menu.inv_alpha;
+	blend_rb = (((menu.rb + bg_rb) >> 8) & 0xFF00FF);
+	blend_g = (((menu.g + bg_g) >> 8) & 0x00FF00);
+	return (blend_rb | blend_g);
+}
+
+void	ft_blend_background(t_img *img, t_img *veil, t_menu menu)
+{
+	char		*img_pixel;
+	char		*veil_pixel;
+	uint32_t	blend_colour;
+	t_vec2i		pos;
+
+	pos.y = -1;
+	while (++pos.y < veil->size.y)
+	{
+		pos.x = -1;
+		while (++pos.x < veil->size.x)
+		{
+			img_pixel = img->addr
+				+ (pos.y * img->line_len + pos.x * img->bytes);
+			blend_colour = ft_fast_alpha_blend(*(uint32_t *)img_pixel, menu);
+			veil_pixel = veil->addr
+				+ (pos.y * veil->line_len + pos.x * veil->bytes);
+			*(uint32_t *)veil_pixel = blend_colour;
+		}
+	}
+}
+
+void	ft_put_pix_to_image(t_img *img, int x, int y, int color)
+{
+	int		i;
+	char	*pixel;
+
+	i = img->bpp - 8;
+	pixel = img->addr + (y * img->line_len + x * img->bytes);
+	while (i >= 0)
+	{
+		if (img->endian != 0)
+			*pixel++ = (color >> i) & 0xFF;
+		else
+			*pixel++ = (color >> (img->bpp - 8 - i)) & 0xFF;
+		i -= 8;
+	}
+}
+
+void	ft_render_image(t_render *render)
+{
+	int			iterate[2];
+	t_colour	pixel_colour;
+	int			colour;
+
+	iterate[1] = -1;
+	while (++iterate[1] < render->cam.image.y)
+	{
+		iterate[0] = -1;
+		while (++iterate[0] < render->cam.image.x)
+		{
+			pixel_colour = ft_anti_aliase_colour(iterate, \
```

Comment:
We can set the number of samples to 1. It is not exactly the same as turning it off as the ray would still get randomly offset, but the difference would be barely visible to the human eye

Reply:
I think it would be a noticeable speed difference.
It could be a bool which chooses the function.

In this context: why even have random samples and not fixed ones (up, down, left, right), if one random sample would effectively the same as other function? Would increase speed since no random calculations.

Reply:
Cause random samples are cool. Of course it costs time. But it's cool.

I mean I can implement a simple bool but we'd had to run it with a performance tester to see if it rly does make a difference

Reply:
If the difference in the final picture is negligible, imo it would be more cool if we render faster.

Right now we have to wait pretty long for one bonus render. Multithreading can speed up, but not that much. 

We can run a performance tester (perf is my choice in makefile, since it doesn't need recompile) but just simple math: 
- pix_centre - cam->centre 
- generate random_float; add to pix_centre; pix_centre - cam->centre

Reply:
Some data:
- Call ft_render_image directly from main (not entering mlx loop)
- One with base version; other with bonus (anti_alias, SAMPLE_SIZE set to 1)
- 3 runs, run with perf

Base: 4220,485 ms (average)
Anti-alias: 5561,727 ms (average)

So about 1,3 sec difference

Length of function calls is a bit misleading since different stuff happens.

As you predicted, the difference is neglible
![comparison](https://github.com/u413-284-si/miniRT/assets/94933640/ff4cb389-64e7-4817-9570-52792e379902)

I can add bool to switch between versions.

Reply:
Good testing, quite insightful. Let's discuss together how you imagine that switch